### ISSUE TITLE
Define how we deal with breaking changes

### ISSUE THREAD

**Issue by mpscholten** [digitallyinduced/ihp#489]
Currently we release a new IHP version every two weeks. When we do breaking changes every two weeks, it makes IHP projects hard to use as you need to spend time every two weeks upgrading.

When it's just a change in the `default.nix` that's easy. But e.g. the last release wasn't so easy to update as there was lot's of trouble with haskell language server and the related nixpkgs upgrade.

Maybe we should batch together the changes that need some work to upgrade. Then we could release small tweaks and bug fixes every two weeks and a big release every 6 weeks.

**Comment by fegu** [comment]
Yes, I agree. Alternating breaking and non-breaking at some frequency is a good idea. But most important is a documented upgrade path for those coming back after a while. Let's say I've done a project using IHP and had it in production for a while. When I want to tweak it I probably want to upgrade first. What I want to know is: do I need to go through all releases one by one (26 in a year, that is a lot of work), can I go directly to the latest one, or is there a smaller set of checkpoints that I need to step by (i.e. you need to upgrade to march 3rd, then october 7th, then latest). Especially, I want to know what changes I have to make to my own code modules to (if any). One document UPGRADING.md should be there in the root folder telling me how. For most active users (at least soon) IHP is not the pet project, they won't follow releases closely, we need to not push these people on older releases away, but make them come back for more.

**Comment by mpscholten** [comment]
Very good points. The `UPGRADING.md` is a good solution.

We could use version numbers like `major.minor.patch`. Whenever `patch` changes this only means an update to the version in the `default.nix`. These can be skipped over easily.

E.g.:

```
v20201016 -- current
v0.2.0 -- next release: just need to change default.nix
v0.2.1 -- change commit in default.nix
v0.2.2 -- change commit in default.nix
v0.2.3 -- change commit in default.nix
v0.3.0 -- breaking change that requires manual steps besides changing default.nix
v0.3.1 -- change commit in default.nix
```

This means users can easily switch from `v0.2.0` to `v0.3.0` without needing to update to `v0.2.1`, `v0.2.2` and `v0.2.3`.

`patch` can include new features as long as they don't break anything.

At the end of the year (maybe already at the beginning of dec) we'll drop the beta label and then change the release cycle from every two weeks to once a month.

**Comment by fegu** [comment]
Yes, this versioning is generally accepted. But keeping the UPGRADING doc, well, updated, is key.

**Comment by HugoPeters1024** [comment]
I do think that at some point we want to guarantee backwards compatibility until some major version release
